pkg/migration-tool/reader/reader.go

```diff
@@ -0,0 +1,174 @@
+	"github.com/timescale/promscale/pkg/log"
+	plan "github.com/timescale/promscale/pkg/migration-tool/planner"
+)
+
+const (
+	megaByte                  = 1024 * 1024
+	ResponseDataSizeHalfLimit = megaByte * 25
+	DefaultReadTimeout        = time.Minute * 5
+	MaxTimeRangeDeltaLimit    = time.Minute * 32
+)
+
+type RemoteRead struct {
+	url             string
+	client          remote.ReadClient
+	plan            *plan.Plan
+	fetchersRunning atomic.Uint32
+	sigBlockWrite   chan struct{}
+	sigBlockRead    chan struct{} // To the writer.
+}
+
+// New creates a new RemoteRead. It creates a ReadClient that is imported from Prometheus remote storage.
+// RemoteRead takes help of plan to understand how to create fetchers.
+func New(readStorageUrl string, p *plan.Plan, sigRead, sigWrite chan struct{}) (*RemoteRead, error) {
+	parsedUrl, err := url.Parse(readStorageUrl)
+	if err != nil {
+		return nil, fmt.Errorf("url-parse: %w", err)
+	}
+	rc, err := remote.NewReadClient(fmt.Sprintf("reader-%d", 1), &remote.ClientConfig{
+		URL:     &config.URL{URL: parsedUrl},
+		Timeout: model.Duration(DefaultReadTimeout),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating remote-client: %w", err)
+	}
+	read := &RemoteRead{
+		url:           readStorageUrl,
+		plan:          p,
+		client:        rc,
+		sigBlockRead:  sigRead,
+		sigBlockWrite: sigWrite,
+	}
+	read.fetchersRunning.Store(0)
+	return read, nil
+}
+
+// Run runs the remote read and starts fetching the samples from the read storage.
+func (rr *RemoteRead) Run(wg *sync.WaitGroup, readerUp *atomic.Bool) error {
+	defer func() {
+		close(rr.sigBlockRead)
+		close(rr.sigBlockWrite)
+		readerUp.Store(false)
+		wg.Done()
+	}()
+	readerUp.Store(true)
+	log.Info("msg", "reader is up")
+	timeRangeMinutesDelta := time.Minute.Milliseconds()
+	for i := rr.plan.Mint; i <= rr.plan.Maxt; {
+		var (
+			mint     = i
+			maxt     = mint + timeRangeMinutesDelta
+			fetch    = rr.createFetch(rr.url, mint, maxt)
+			blockRef = rr.plan.CreateBlock(mint, maxt)
+		)
+		blockRef.SetDescription(fmt.Sprintf("fetching time-range: %d mins...", timeRangeMinutesDelta/time.Minute.Milliseconds()), 1)
+		result, err := fetch.start(context.Background(), []*labels.Matcher{labels.MustNewMatcher(labels.MatchRegexp, "", ".*")})
+		if err != nil {
+			return fmt.Errorf("remote-read run: %w", err)
+		}
+		if len(result.Timeseries) == 0 {
+			rr.plan.DecrementBlockCount()
+			i += timeRangeMinutesDelta + 1
+			if timeRangeMinutesDelta < MaxTimeRangeDeltaLimit.Milliseconds() {
+				timeRangeMinutesDelta *= 2
+			}
+			continue
+		}
+		qResultSize := result.Size() // Assuming each character corresponds to one byte.
+		rr.plan.UpdateLastPushedMaxt(i)
+		blockRef.SetDescription(fmt.Sprintf("received %d MB with delta %d mins...", qResultSize/megaByte, timeRangeMinutesDelta/time.Minute.Milliseconds()), 1)
+		blockRef.Timeseries = result.Timeseries
+		rr.sigBlockRead <- struct{}{}
+		<-rr.sigBlockWrite
```

Comment:
go channels auto block if they reached capacity. sigBlockWrite can be removed, and if the writer hasn't processed the previous signal then `rr.sigBlockRead <- struct{}{} ` will block

Reply:
Yes I agree, but the lib I am using for progress-bar has some issues with concurrency, so I have to be extra careful that at a single time, only one block write to progress bar. If I do that way, it will to 2 blocks writing, one by the writer and the other one waiting in memory for the next write iteration.

Reply:
Sorry, this `some issues with concurrency` went wrong. I meant to say that that lib has issues when two of its instance writes to the stdout, it shows weird behavior. However, I have figured out another library that handles multi-bars pretty well. I will update to that soon. Hence, not resolving this as it's still a TODO.

Reply:
Also, each write on a progress-bar is concurrency safe. The issue is only when more than one bar are written simultaneously to the stdout.